Bucket Notifications Unification

welansari opened the pull request:
- Added support for specifying multiple destination kafka hosts in bucket notification (ported logic from 7.x)
- Ported NotificationConfigManager from 7.x with tests
- Added an abstract NotificationConfigManager class to support both MongoDB and Zookeeper backends
- Ported code from 7.x that handles updating notification configs in the Zookeeper backend
- NotificationConfigManager now returns 7.x config structure, this was changed to keep the code functional with previously stored configs in Zookeeper on S3C.
- Support S3C delete notifications. 
- Made metastore collection name configurable as the name is different between S3c and Zenko.
- Fixed config not getting removed from zookeeper when a bucket is deleted.

**Note:** To ease reviews, please review commit by commit.

Issue: BB-497

Commit:
support specifying multiple destination kafka hosts in bucket notification

S3C supports passing multiple Kafka hosts in the host field under the format
"host1:port1,host2:port2"

Issue: BB-497

Commit:
port NotificationConfigManager from 7.x

7.x's NotificationConfigManager uses Zookeeper to persitently
store bucket notification configurations.

Issue: BB-497

Commit:
abstract NotificationConfigManager to support both MongoDB and Zookeeper

Added an abstraction class to allow using the NotificationConfigManager
with both MongoDB and Zookeeper.

Issue: BB-497

Commit:
use 7.x bucket notification config structure

Using 7.x bucket notification config structure as the config
is persisted in Zookeeper in S3C and will need to be migrated if we decide
to keep 8.x format which returns the config directly without the bucket
name.

Issue: BB-497

Commit:
add/delete notification config when using zookeeper backend

ported code from 7.x

Issue: BB-497

Commit:
support notifications of S3C delete events

S3C uses the real delete oplog event contrary to Zenko
which uses the update event we do before each delete.

Delete events don't contain any object metadata.

The key of the object is extracted from the document's ID
which also contains the decoded versionID.

The encoded versionID is extracted from the overhead fields
attached to the event log.

The event dateTime used is the Kafka commit timestamp.

"ignoreEmptyEvents" was added to the config to avoid having double
delete notifications as Zenko shouldn't use true delete events.

Issue: BB-497

bert-e commented:
# Hello kerkesni,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/backbeat/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-497 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.6.50`

* `8.7.0`


Please check the `Fix Version/s` of BB-497, or the target
branch of this pull request.

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

Commit:
make metastore collection name configurable for bucket notifications

Issue: BB-497

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-497 contains:


* `8.6.50`

* `8.7.0`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.7.0`


Please check the `Fix Version/s` of BB-497, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

Commit:
fix config not getting removed from zookeeper when bucket is deleted

Issue: BB-497

Commit:
add tests

Issue: BB-497

francoisferrand commented on extensions/lifecycle/LifecycleQueuePopulator.js:
```diff
@@ -490,6 +490,10 @@ class LifecycleQueuePopulator extends QueuePopulatorExtension {
      * @return {undefined}
      */
     filter(entry) {
+        if (!entry.value) {
```
can't this happen with `delete` entries?

welansari replied:
We used to ignore these in the logReader, i moved that logic to the actual extensions as S3c uses true delete events for bucket notifications. In lifecycle we use the update we do before each delete as we need the metadata.

francoisferrand commented on extensions/notification/NotificationConfigManager.js:
```diff
@@ -1,283 +1,87 @@
-            this._logger.debug('Connected to MongoDB', {
-                method: 'NotificationConfigManager._setupMongoClient',
+    constructor(params) {
+        this._usesZookeeperBackend = false;
```
seems like an anti-pattern: we already have abstracted away zookeeper/mongo through `_configManagerBackend`, so either
* this needs to be moved to the backend
* every backend needs to have a new method, called at the right time, to trigger this zookeeper-specific handling

francoisferrand commented on extensions/notification/NotificationConfigManager.js:
```diff
@@ -1,283 +1,87 @@
-            method: 'NotificationConfigManager._handleChangeStreamChange',
-        });
+        if (val instanceof Promise) {
+            return val.then(res => cb(null, res)).catch(err => cb(err));
```
not needed or not tested...

francoisferrand commented on extensions/notification/NotificationConfigManager.js:
```diff
@@ -1,283 +1,87 @@
-        // start watching metastore
-        this._metastoreChangeStream.start();
+    setConfig(bucket, config) {
+        if (this._usesZookeeperBackend) {
```
not needed : implement the method in mongo backend as well, and just return false

francoisferrand commented on extensions/notification/NotificationConfigManager.js:
```diff
@@ -1,283 +1,87 @@
-            return cb();
-        });
+    removeConfig(bucket) {
+        if (this._usesZookeeperBackend) {
```
not needed : implement the method in mongo backend as well, and just return false

francoisferrand commented on extensions/notification/NotificationQueuePopulator.js:
```diff
@@ -200,24 +312,29 @@ class NotificationQueuePopulator extends QueuePopulatorExtension {
     filterAsync(entry, cb) {
-        const { bucket, key, overheadFields } = entry;
+        if (this.notificationConfig.ignoreEmptyEvents && !entry.value) {
+            return cb();
```
why do we we have a config `ignoreEmptyEvents` ?
- either this can happen & there is nothing to do → ignore them
- or it can happen & but there is nothing to do → ignore them
- or it can happen & there is something to do → we must not ignore them...

welansari replied:
Empty events have to be ignored in Zenko, else we'll get duplicate delete notifications. We use the update we do before each delete for delete notifications.
In S3C empty delete events are the only event available for delete notifications.

francoisferrand replied:
is this filtering (in zenko) not already done in the mongo abstraction layers (e.g. either MongoClientInterface or LogReader), where we also do generate a "delete" event from an update with "deleted=true" ?

It seems this would really be the proper place to handle it, keeping proper layering & abstractions.

francoisferrand replied:
Reading further, I see you removed this from `LogReader` : it seems to me this should not be "moved up" to QueuePopulator/..., but instead "moved down" into the specific implementations of `LogReader` (i.e. maybe a virtual function in LogReader, defined differently in each implementation ; or a flag from LogReader initialized by each subclass...)

francoisferrand commented on extensions/notification/configManager/MongoConfigManager.js:
```diff
@@ -0,0 +1,291 @@
+
+function onConfigManagerConfigGet(cacheHit, delay) {
+    configGetLag.observe({
+        [CONFIG_MANAGER_CACHE_HIT]: cacheHit,
```
```suggestion
        CONFIG_MANAGER_CACHE_HIT: cacheHit,
```

welansari replied:
We have to use the brackets else the label literally becomes `CONFIG_MANAGER_CACHE_HIT `

francoisferrand commented on extensions/notification/configManager/MongoConfigManager.js:
```diff
@@ -0,0 +1,291 @@
+
+function onConfigManagerCacheUpdate(op) {
+    cacheUpdates.inc({
+        [CONFIG_MANAGER_OPERATION_TYPE]: op,
```
```suggestion
        CONFIG_MANAGER_OPERATION_TYPE: op,
```

welansari replied:
We have to use the brackets else the label literally becomes `CONFIG_MANAGER_OPERATION_TYPE`

francoisferrand commented on extensions/notification/queueProcessor/QueueProcessor.js:
```diff
@@ -141,6 +165,7 @@ class QueueProcessor extends EventEmitter {
      */
     start(options, done) {
         async.series([
+            next => this._setupZookeeper(next),
```
can't we setup zookeper inside the `ZookeeperConfigManager` ?

welansari replied:
The queue populator creates a zookeeper client for all queue populator extensions, we kind of have to do it this way in the queue processor.

francoisferrand replied:
I don't understand why.
* this specific attribute (`zkClient`) is initialized here, once for every call to `start()`
* it seems it is only used on the next function (`_setupNotificationConfigManager`)
* this function does not store zkClient, it just passes it to notificationManager constructor, which passes it only to the zookeerConfigManager...

--> It seems to me we may as well initialize the zkClient inside `zookeeperConfigManager` constructor (and avoid an extra zkClient in Zenko context) 

(but we do indeed need to pass zkConfig)

francoisferrand replied:
ok, i see what you mean: issue is not here, but in QueuePopulator where we want to reuse the one we already have (even in Zenko).

- However, this current implementation creates a useless zkClient in QueueProcessor on Zenko : which should be avoided.
- One way to avoid this may be to pass both zkConfig and an (optional) zkClient to `NotificationConfigManager` : so that if provided with a client, it can use it (QueuePopulator case) but it will otherwise create one from the config (QueueProcessor case).

francoisferrand commented on extensions/notification/utils/config.js:
```diff
@@ -87,9 +87,14 @@ function validateEntry(bnConfig, entry) {
     }
 
+    if (bucket !== bnConfig.bucket) {
+        return { isValid: false, matchingConfig: null };
```
should be tested

francoisferrand commented on extensions/notification/queueProcessor/QueueProcessor.js:
```diff
@@ -84,14 +86,34 @@ class QueueProcessor extends EventEmitter {
+            return;
+        }
+
+        const populatorZkPath = this.notifConfig.zookeeperPath;
```
can we add some tests for this file?

francoisferrand commented on extensions/notification/NotificationConfigManager.js:
```diff
@@ -1,283 +1,79 @@
-     * @param {Function} cb callback
-     * @returns {undefined}
+     * Remove bucket notification configuration
+     * Only supported in the Zookeeper backend.
```
not really true : we have implemented an abstraction, the method is supported on every backend... (even if some backends actually have an _empty_ implementation)

francoisferrand commented on extensions/notification/NotificationConfigManager.js:
```diff
@@ -1,283 +1,79 @@
-     * partitions and so don't need the configs for all buckets
-     * @returns {undefined}
+     * Add/update bucket notification configuration.
+     * Only supported in the Zookeeper backend.
```
not really true : we have implemented an abstraction, the method is supported on every backend... (even if some backends actually have an empty implementation)

francoisferrand commented on extensions/notification/configManager/MongoConfigManager.js:
```diff
@@ -0,0 +1,311 @@
+
+    /**
+     * Set bucket notification configuration
+     * (Not implemted in the MongoDB backend)
```
it is not "not implemented", more like it is not "needed" : because mongo backend automatically configures itself (using a change stream)

francoisferrand commented on extensions/notification/queueProcessor/QueueProcessor.js:
```diff
@@ -240,19 +262,15 @@ class QueueProcessor extends EventEmitter {
                         // skip, if there is no config for the current
                         // destination resource
-                        return done();
+                        return undefined;
```
should this not call `done()` as before?

francoisferrand commented on lib/queuePopulator/QueuePopulator.js:
```diff
@@ -417,6 +417,8 @@ class QueuePopulator {
                 this.bnConfigManager = new NotificationConfigManager({
                     mongoConfig: this.qpConfig.mongo,
+                    bucketMetastore: this.extConfigs.notification.bucketMetastore,
+                    zkClient: this.qpConfig.mongo ? null : this.zkClient,
```
why filter the parameters?
NotificationConfigManager takes multiple optional params (mongo, zkClient...) and deals with it as needed, no point doing it for him here.

anurag4DSB commented on extensions/notification/configManager/ZookeeperConfigManager.js:
```diff
@@ -0,0 +1,446 @@
+        return result;
+    }
+
+    _getBucketNotifConfig(bucket, cb) {
```
its worth adding some logs here for easier debugging.
Same for methods, `_removeBucketNotifConfigNode`, `_setBucketNotifConfig`

anurag4DSB commented on extensions/notification/configManager/ZookeeperConfigManager.js:
```diff
@@ -0,0 +1,446 @@
+    }
+
+    _updateLocalStore(buckets, cb) {
+        async.eachSeries(buckets, (bucket, next) => {
```
have you thought about using  `async.eachLimit` with a configurable limit instead of `eachSeries` in `_updateLocalStore` for improved performance on large datasets(controlled parallel processing of tasks)

anurag4DSB commented on extensions/notification/configManager/ZookeeperConfigManager.js:
```diff
@@ -0,0 +1,446 @@
+            bucket,
+            config,
+        });
+        this._setBucketNotifConfig(bucket, JSON.stringify(config), err => {
```
Should we add a retry logic to handle transient network issues?
same for `_getBucketNotifConfig`

welansari replied:
The Zookeeper library we use already handles retries on network issues. Only needed to enable it by configuring the max number of retries.

anurag4DSB commented on extensions/notification/configManager/MongoConfigManager.js:
```diff
@@ -0,0 +1,312 @@
+                bucket,
+                error: err.message,
+            });
+            return undefined;
```
should we return the err here?

anurag4DSB commented on extensions/notification/configManager/MongoConfigManager.js:
```diff
@@ -0,0 +1,312 @@
+     * @param {String} bucket - bucket
+     * @return {Object|undefined} - configuration if available or undefined
+     */
+    async getConfig(bucket) {
```
can we add some debug logs for cache hit-and-miss scenarios to improve observability of cache performance and usage

welansari replied:
We could use this metric for that `s3_notification_config_manager_config_get_seconds`. We also have a metric for how many times we update the cache `s3_notification_config_manager_cache_updates_total`

anurag4DSB commented on extensions/notification/configManager/MongoConfigManager.js:
```diff
@@ -0,0 +1,312 @@
+    logger: joi.object().required(),
+}).required();
+
+const MAX_CACHED_ENTRIES = Number(process.env.MAX_CACHED_BUCKET_NOTIFICATION_CONFIGS)
```
can this be a config item as well?

anurag4DSB commented on extensions/notification/NotificationQueuePopulator.js:
```diff
@@ -42,6 +42,75 @@ class NotificationQueuePopulator extends QueuePopulatorExtension {
+     * @return {Object|undefined} - bucket attributes if available
+     */
+    _getBucketAttributes(value) {
+        if (value && value.attributes) {
```
should we cache this in `this._bucketAttributes` or something?(Lazy Initialization of _getBucketAttributes)

anurag4DSB commented on extensions/notification/NotificationQueuePopulator.js:
```diff
@@ -42,6 +42,75 @@ class NotificationQueuePopulator extends QueuePopulatorExtension {
+            };
+            // bucket notification config is available, update node
+            this.bnConfigManager.setConfig(bucketName, bnConfig);
+            return undefined;
```
is it possible to keep the return consistent with the below, either returning the defined below or changing here?

anurag4DSB commented on extensions/notification/NotificationQueuePopulator.js:
```diff
@@ -3,6 +3,7 @@ const util = require('util');
 
 const { isMasterKey } = require('arsenal').versioning;
 const { usersBucket, mpuBucketPrefix, supportedNotificationEvents } = require('arsenal').constants;
+const VID_SEP = require('arsenal').versioning.VersioningConstants.VersionId.Separator;
```
NIT: Worth writing `VID_SEPERATOR`, I am okay with VID :)

anurag4DSB reviewed (commented):
I added some comments; I am curious how we are testing this code as it includes s3c-specific changes.

Commit:
retry zookeeper operations on transient network issues

The zookeeper lib we use can handle connection loss errors,
but retries are not configured when intializing the client.
Default is no retry.

Issue: BB-497

Commit:
make max cached notification configs a config prameter

Issue: BB-497

Commit:
concurrently update local bucket notification configs from zookeeper

Issue: BB-497

Commit:
limit the number of event handlers on notification zookeeper events

The number of event handlers we have for each event is controlled.
No need to remove the limit of how many we can have as it adds risk
of having a memory leak.

Issue: BB-497

Commit:
add more logs in the zookeeper config manager

Issue: BB-497

Commit:
Throw when getConfig fails

Issue: BB-497

Commit:
only parse bucket entries once

Issue: BB-497

welansari commented:
> I am curious how we are testing this code as it includes s3c-specific changes

@anurag4DSB I'm manually testing everything on a RING with custom images and modified federation steps to have the proper configs for Backbeat.

francoisferrand commented on extensions/notification/configManager/MongoConfigManager.js:
```diff
@@ -0,0 +1,312 @@
+        if (cachedConfig) {
+            const delay = (Date.now() - startTime) / 1000;
+            onConfigManagerConfigGet(true, delay);
+            return {
```
what is the point of returning the `bucket`?
it seems we are not using it (kind of logical, since we already have it to get the config...)

welansari replied:
The old config manager used to store the config in zookeeper this way. To avoid having to migrate the old data stored in zookeeper or having to deal with the old vs new format i decided to keep the old one.

francoisferrand commented on extensions/notification/configManager/MongoConfigManager.js:
```diff
@@ -0,0 +1,312 @@
+ * are directly retrieved from the metastore, and are locally cached. Cache
+ * is invalidated using MongoDB change streams.
+ */
+class MongoConfigManager {
```
it seems to me we are missing a "base" class, if only to document the interface these must conform to.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-497 contains:


* `8.7.0`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `9.0.0`


Please check the `Fix Version/s` of BB-497, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

welansari commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/9.0`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.5`
* `development/8.6`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/9.0`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.5`
* `development/8.6`



Please check the status of the associated issue BB-497.

Goodbye kerkesni.
